TaskLayer/MetaMorpheusTask.cs

```diff
@@ -598,25 +602,120 @@ private static bool SameSettings(string pathToOldParamsFile, IndexingEngine inde
             return false;
         }
 
-        private static void WritePeptideIndex(List<PeptideWithSetModifications> peptideIndex, string peptideIndexFile)
+        private static void WritePeptideIndex(List<PeptideWithSetModifications> peptideIndex, List<Protein> proteins, string peptideIndexFile, int maxThreads)
         {
-            var messageTypes = GetSubclassesAndItself(typeof(List<PeptideWithSetModifications>));
+            List<string> allKnownMods = GlobalVariables.AllModsKnown.Select(x => x.IdWithMotif).ToList();
+            List<string> allKnownProteins = proteins.Select(x => x.Accession).ToList();
+
+            List<int>[] indexToWrite = new List<int>[peptideIndex.Count];
+            int[] threads = Enumerable.Range(0, maxThreads).ToArray();
+            int peptidesPerThread = peptideIndex.Count / maxThreads + 1; //+1 more equitably distributes the remainder so the last thread doesn't do heavier lifting
+            Parallel.ForEach(threads, (thread) =>
+            {
+                int i = thread * peptidesPerThread;
+                int endIndex = thread == maxThreads - 1 ?
+                    peptideIndex.Count :
+                    (thread + 1) * peptidesPerThread;
+
+                for (; i < endIndex; i++)
+                {
+                    PeptideWithSetModifications peptide = peptideIndex[i];
+
+                    //convert the peptide into a list of integers
+                    //Making a choice here to remove the peptide description, 
+                    //because integers are 4x faster to serialize than strings 
+                    //and there's no clean conversion for the description
+                    List<int> indexedPeptide = new List<int>
+                    {
+                        allKnownProteins.IndexOf(peptide.Protein.Accession), //0
+                        peptide.OneBasedStartResidueInProtein, //1
+                        peptide.OneBasedEndResidueInProtein, //2
+                        Convert.ToInt16(peptide.CleavageSpecificityForFdrCategory), //3
+                        peptide.MissedCleavages, //4
+                        peptide.NumFixedMods //5
+                    };
+
+                    //add on the modifications
+                    foreach (KeyValuePair<int, Modification> kvp in peptide.AllModsOneIsNterminus)
+                    {
+                        indexedPeptide.Add(kvp.Key);  //6, 8, 10, etc
+                        string idWithMotif = kvp.Value.IdWithMotif;
+                        indexedPeptide.Add(allKnownMods.IndexOf(idWithMotif)); //7, 9, 11, etc
```

Comment:
I included a sort for the proteins (because they're not always read in the same), but changing the mods is problematic. Not sure who would do that, but alright. I don't see a way around fixing that, so we can discard this PR.